src/libexpr-c/nix_api_expr.h

```diff
@@ -129,8 +129,7 @@ nix_err nix_value_call_multi(
  *
  * This function converts these Values into their final type.
  *
- * @note You do not need this function as long as you only use API calls that return a `Value`. This function is mainly
- * needed before calling @ref getters.
+ * @note This function is called before calling `nix_get_type()` (and therefore before most other @ref getters), because lazy evaluation should be transparent to users.
```

Comment:
```suggestion
 * @note This function should be called before calling `nix_get_type()` (and therefore before most other @ref getters), because lazy evaluation should be transparent to users.
```

Otherwise it sounds a bit like we're describing what the bindings do. It is the reader's responsibility.

Reply:
```suggestion
 * @note You only need to call this function before `nix_get_type()` (and therefore before most other @ref getters), because lazy evaluation should be transparent to users of this API.
```

How about this then?

Reply:
That's not necessarily true. Suppose in your application you have a hardcoded Nix function, defined in a string literal, so you know how it behaves. For example:

```nix
# passed to nix_expr_eval_from_string

{ foo, bar }:
let
  # fancy application-specific computations are declared here
in
{
  inherit baz;
  n = builtins.length qux;
}
```

In such a case it's valid to

1. call the function
2. retrieve the `n` `Value`
3. `nix_value_force` `n`
4. _not_ check the type, because we know its type statically
5. call `nix_get_int`

Obviously a toy example, but this is a useful implementation pattern when you have a situation where a few things related things need to be computed by Nix and the application then reads the values it needs. Unlike individual function calls, you take advantage of shared computations this way.

Reply:
Okay, maybe I'm misunderstanding something big time here. Apparently it's not *required* to call it beforehand, but sometimes it is? But saying "you should" is not very informative - do you have to or not, what are the considerations exactly?

Reply:
i thought, "you need `nix_value_force` before any api function if and only if that api function does not return a `Value`" was a beautifully simple rule. is it wrong? will https://github.com/NixOS/nix/issues/10499 make it wrong? cc @roberth